File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Action/Common/ReducePowerAction.ts
import Async from "../../../../../Core/Async";
import { AbstractPower } from "../../Power/AbstractPower";
import AbstractCreature from "../../Unit/AbstractCreature";
import AbstractGameAction from "../AbstractGameAction";

export class ReducePowerAction extends AbstractGameAction{
    public constructor(public Target:AbstractCreature,public Source:AbstractCreature,public power:string|AbstractPower,public Amount:number) {
        super();
    }

    async Update(): Promise<boolean> {
        if(this.Duration == 0){
            let reduceMe:AbstractPower;
            if(this.power){
                if(typeof this.power === 'string'){
                    reduceMe = this.Target.GetPower(this.power);
                }
                else{
                    reduceMe = <AbstractPower>this.power;
                }
                if(reduceMe){
                    reduceMe.Amount -= this.Amount;
                }
            }
        }
        await Async.Delay(0.25);
        return true;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Action/Utility/ReducePowerAction.cs
using Cysharp.Threading.Tasks;
using ModTheSpire.Event;
using ModTheSpire.Power;
using ModTheSpire.Unit;

namespace ModTheSpire.Effect.Other
{
    public class ReducePowerAction : IGameAction
    {
        public float Duration { get; set; }
        public ActionType Type { get; } = ActionType.POWER;
        
        private AbstractCreature Target;
        private AbstractCreature Source;
        private AbstractPower Power;
        private int Amount;

        public ReducePowerAction(AbstractCreature target, AbstractCreature source, AbstractPower power,int amount)
        {
            Target = target;
            Source = source;
            Power = power;
            Amount = amount;
            Power = this.Target.GetPower(power?.Id);
        }

        public ReducePowerAction(AbstractCreature target, AbstractCreature source, string power,int amount)
        {
            Target = target;
            Source = source;
            Amount = amount;
            Power = this.Target.GetPower(power);
        }
        public void Init()
        {
        }

        public async UniTask<bool> Update()
        {
            if (Power != null)
            {
                Power.Amount -= this.Amount;
            }

            await UniTask.Delay(250);
            return true;
        }
    }
}